scipy/fft/tests/test_basic.py

```diff
@@ -0,0 +1,513 @@
+import queue
+import threading
+import multiprocessing
+import numpy as np
+import pytest
+from numpy.random import random
+from numpy.testing import (
+    assert_array_almost_equal, assert_array_equal, assert_allclose
+)
+from pytest import raises as assert_raises
+import scipy.fft as fft
+from scipy.conftest import (
+    array_api_compatible,
+    skip_if_array_api,
+    skip_if_array_api_gpu,
+    set_assert_allclose
+)
+from scipy._lib._array_api import _assert_matching_namespace, array_namespace, size
+
+
+def fft1(x):
+    L = len(x)
+    phase = -2j*np.pi*(np.arange(L)/float(L))
+    phase = np.arange(L).reshape(-1, 1) * phase
+    return np.sum(x*np.exp(phase), axis=1)
+
+
+class TestFFTShift:
+
+    @array_api_compatible
+    def test_fft_n(self, xp):
+        x = xp.asarray([1, 2, 3])
+        if xp.__name__ in 'torch':
+            assert_raises(RuntimeError, fft.fft, x, 0)
+        else:
+            assert_raises(ValueError, fft.fft, x, 0)
+
+
+class TestFFT1D:
+
+    @array_api_compatible
+    def test_identity(self, xp):
+        maxlen = 512
+        x = xp.asarray(random(maxlen) + 1j*random(maxlen))
+        xr = xp.asarray(random(maxlen))
+        _assert_allclose = set_assert_allclose(xp)
+        for i in range(1, maxlen):
+            _assert_allclose(fft.ifft(fft.fft(x[0:i])), x[0:i], rtol=1e-12, atol=0)
+            _assert_allclose(
+                fft.irfft(fft.rfft(xr[0:i]), i), xr[0:i], rtol=1e-12, atol=0)
```

Comment:
These were previously `assert_array_almost_equal` with `decimal=12`. They usually pass, but I have seen them occasionally fail to meet the precision required by `rtol=1e-12`  - by 1 element out of hundreds - on the `numpy` backend.